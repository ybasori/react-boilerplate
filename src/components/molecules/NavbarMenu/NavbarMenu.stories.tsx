import type { Meta, StoryObj } from "@storybook/react";
import NavbarMenu from "./NavbarMenu";
import { MemoryRouter } from "react-router-dom";
import { fn } from "@storybook/test";

//ðŸ‘‡ This default export determines where your story goes in the story list
const meta: Meta<typeof NavbarMenu> = {
  title: "Molecules/NavbarMenu",
  component: (props) => (
    <MemoryRouter>
      <NavbarMenu {...props} />
    </MemoryRouter>
  ),
};

export default meta;
type Story = StoryObj<typeof NavbarMenu>;

export const Default: Story = {
  args: {
    menu: [
      {
        to: "/test",
        name: "Test",
      },
      {
        onClick: fn(),
        name: "Click Test",
      },
    ],
  },
};
