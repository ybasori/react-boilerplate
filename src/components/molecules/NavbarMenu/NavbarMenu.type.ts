export interface IProps {
  menu: { to?: string; name: string; onClick?: () => void }[];
}
