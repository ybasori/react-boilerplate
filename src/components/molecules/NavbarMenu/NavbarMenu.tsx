import { Fragment } from "react";
import NavbarLink from "@/components/atoms/NavbarLink/NavbarLink";
import { IProps } from "./NavbarMenu.type";
import NavbarButton from "@/components/atoms/NavbarButton/NavbarButton";

const NavbarMenu: React.FC<IProps> = ({ menu }) => {
  return (
    <>
      <ul>
        {menu.map((item, key) => (
          <Fragment key={key}>
            {!!!item.onClick && !!item.to ? (
              <NavbarLink to={item.to} name={item.name} />
            ) : (
              <NavbarButton
                onClick={item.onClick ?? (() => null)}
                name={item.name}
              />
            )}
          </Fragment>
        ))}
      </ul>
    </>
  );
};

export default NavbarMenu;
