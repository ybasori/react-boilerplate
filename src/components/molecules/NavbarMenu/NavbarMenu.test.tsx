import { fireEvent, render, screen } from "@testing-library/react";
import NavbarMenu from "./NavbarMenu";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

describe("Molecules/NavbarMenu", () => {
  test("renders the NavbarMenu with the correct option title", () => {
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <NavbarMenu
          menu={[
            {
              name: "menu",
            },
            {
              name: "menu two",
            },
          ]}
        />
      </Router>
    );

    // Assert that the button is in the document and has the correct text
    const buttonElement = screen.getByText("menu");
    expect(buttonElement).toBeInTheDocument();
    const buttonElementTwo = screen.getByText("menu two");
    expect(buttonElementTwo).toBeInTheDocument();
  });

  test("calls the onClick handler when the menu is clicked", () => {
    const history = createMemoryHistory();

    const handleClick = jest.fn();
    render(
      <Router history={history}>
        <NavbarMenu
          menu={[
            {
              name: "menu",
              onClick: handleClick,
            },
            {
              name: "menu two",
              to: "/test",
            },
          ]}
        />
      </Router>
    );

    const buttonElement = screen.getByText("menu");
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);

    const buttonElementTwo = screen.getByText(`menu two`);
    fireEvent.click(buttonElementTwo);
    expect(history.location.pathname).toBe("/test");
  });

  test("Snapshot", () => {
    const history = createMemoryHistory();

    const handleClick = jest.fn();
    const { container } = render(
      <Router history={history}>
        <NavbarMenu
          menu={[
            {
              name: "menu",
              onClick: handleClick,
            },
            {
              name: "menu two",
              to: "/test",
            },
          ]}
        />
      </Router>
    );

    expect(container).toMatchSnapshot();
  });
});
