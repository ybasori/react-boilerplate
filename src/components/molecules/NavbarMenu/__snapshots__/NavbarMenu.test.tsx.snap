// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/NavbarMenu Snapshot 1`] = `
<div>
  <ul>
    <li>
      <span>
        menu
      </span>
    </li>
    <li>
      <a
        href="/test"
      >
        menu two
      </a>
    </li>
  </ul>
</div>
`;
